vmm/src/vm.rs

```diff
@@ -1173,99 +1173,92 @@ impl Vm {
+    fn load_firmware(
+        mut firmware: &File,
+        memory_manager: Arc<Mutex<MemoryManager>>,
+    ) -> Result<EntryPoint> {
         let uefi_flash = memory_manager.lock().as_ref().unwrap().uefi_flash();
         let mem = uefi_flash.memory();
         arch::riscv64::uefi::load_uefi(mem.deref(), arch::layout::UEFI_START, &mut firmware)
             .map_err(Error::UefiLoad)?;
-        Ok(())
+        Ok(EntryPoint {
+            entry_addr: arch::layout::UEFI_START,
+        })
     }
 
     #[cfg(target_arch = "riscv64")]
     fn load_kernel(
-        firmware: Option<File>,
-        kernel: Option<File>,
+        mut kernel: File,
         memory_manager: Arc<Mutex<MemoryManager>>,
     ) -> Result<EntryPoint> {
         let guest_memory = memory_manager.lock().as_ref().unwrap().guest_memory();
         let mem = guest_memory.memory();
         let alignment = 0x20_0000;
         let aligned_kernel_addr = arch::layout::KERNEL_START.0 + (alignment - 1) & !(alignment - 1);
-        let entry_addr = match (firmware, kernel) {
-            (None, Some(mut kernel)) => {
-                match linux_loader::loader::pe::PE::load(
-                    mem.deref(),
-                    Some(GuestAddress(aligned_kernel_addr)),
-                    &mut kernel,
-                    None,
-                ) {
-                    Ok(entry_addr) => entry_addr.kernel_load,
-                    // Try to load the binary as kernel PE file at first.
-                    // If failed, retry to load it as UEFI binary.
-                    // As the UEFI binary is formatless, it must be the last option to try.
-                    Err(linux_loader::loader::Error::Pe(InvalidImageMagicNumber)) => {
-                        Self::load_firmware(&kernel, memory_manager)?;
-                        arch::layout::UEFI_START
-                    }
-                    Err(e) => {
-                        return Err(Error::KernelLoad(e));
-                    }
+        let entry_addr = {
+            match linux_loader::loader::pe::PE::load(
+                mem.deref(),
+                Some(GuestAddress(aligned_kernel_addr)),
+                &mut kernel,
+                None,
+            ) {
+                Ok(entry_addr) => entry_addr.kernel_load,
+                // Try to load the binary as kernel PE file at first.
+                // If failed, retry to load it as UEFI binary.
+                // As the UEFI binary is formatless, it must be the last option to try.
+                Err(linux_loader::loader::Error::Pe(InvalidImageMagicNumber)) => {
+                    Self::load_firmware(&kernel, memory_manager)?;
+                    arch::layout::UEFI_START
+                }
+                Err(e) => {
+                    return Err(Error::KernelLoad(e));
                 }
             }
-            (Some(firmware), None) => {
-                Self::load_firmware(&firmware, memory_manager)?;
```

Comment:
for the record: I replaced this logically with this change:

<img width="1247" height="519" alt="image" src="https://github.com/user-attachments/assets/e5f64459-b8fe-4eff-8bb7-19755cbb59ab" />